File: backends/rust-axum/src/usecase/staff/interactor.rs
//! スタッフユースケース Interactor モジュール。
//!
//! # Author
//! Satoshi Nagashiba <satoshi.nagashiba@gmail.com>

use std::sync::Arc;
use crate::domain::staff::{
    condition::Condition,
    entity::Staff,
    repository::Repository,
    value_objects::ListItem,
};
use super::dto::{UpdateRoleDto, DestroyDto};

pub type UseCaseError = Box<dyn std::error::Error + Send + Sync>;

/// スタッフのユースケース実装。
pub struct Interactor {
    repo: Arc<dyn Repository>,
}

impl Interactor {
    /// リポジトリを受け取りインタラクターを生成します。
    pub fn new(repo: Arc<dyn Repository>) -> Self {
        Self { repo }
    }

    /// 検索条件に合致するスタッフ一覧の VO を返します。
    pub async fn find_by_condition(&self, cond: Condition) -> Result<Vec<ListItem>, UseCaseError> {
        let staffs = self.repo.find_by_condition(cond).await?;
        Ok(staffs.into_iter().map(to_list_item).collect())
    }

    /// スタッフのロールを更新します。
    pub async fn update_role(&self, dto: UpdateRoleDto) -> Result<(), UseCaseError> {
        let ok = self.repo.update_role(dto.id, dto.role, dto.executor_id).await?;
        if !ok {
            return Err("staff_not_found".to_string().into());
        }
        Ok(())
    }

    /// スタッフの論理削除を復元します。
    pub async fn restore(&self, id: u32) -> Result<(), UseCaseError> {
        let ok = self.repo.restore(id).await?;
        if !ok {
            return Err("staff_not_found".to_string().into());
        }
        Ok(())
    }

    /// スタッフを論理削除します。
    pub async fn destroy(&self, dto: DestroyDto) -> Result<(), UseCaseError> {
        let ok = self.repo.soft_delete(dto.id, dto.executor_id).await?;
        if !ok {
            return Err("staff_not_found".to_string().into());
        }
        Ok(())
    }
}

fn to_list_item(s: Staff) -> ListItem {
    let status = if s.deleted_at.is_some() { 0 } else { 1 };
    ListItem {
        id:         s.id,
        name:       s.name,
        email:      s.email,
        role:       s.role,
        status,
        created_at: s.created_at,
        updated_at: s.updated_at,
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use async_trait::async_trait;
    use std::sync::Mutex;
    use crate::domain::staff::{
        condition::Condition,
        entity::Staff,
        repository::{DomainError, Repository},
    };

    struct MockRepo {
        find_by_condition: Mutex<Option<Vec<Staff>>>,
        update_role_ok:    Mutex<bool>,
        restore_ok:        Mutex<bool>,
        soft_delete_ok:    Mutex<bool>,
    }

    impl MockRepo {
        fn new() -> Self {
            Self {
                find_by_condition: Mutex::new(None),
                update_role_ok:    Mutex::new(true),
                restore_ok:        Mutex::new(true),
                soft_delete_ok:    Mutex::new(true),
            }
        }
    }

    fn make_staff(id: u32, deleted: bool) -> Staff {
        let now = chrono::Utc::now();
        Staff {
            id,
            name:          "Test Staff".to_string(),
            email:         "staff@example.com".to_string(),
            provider:      1,
            provider_id:   "google123".to_string(),
            avatar:        None,
            role:          2,
            last_login_at: None,
            created_at:    now,
            created_by:    None,
            updated_at:    now,
            updated_by:    None,
            deleted_at:    if deleted { Some(now) } else { None },
            deleted_by:    None,
            version:       0,
        }
    }

    #[async_trait]
    impl Repository for MockRepo {
        async fn find_by_condition(&self, _cond: Condition) -> Result<Vec<Staff>, DomainError> {
            Ok(self.find_by_condition.lock().unwrap().take().unwrap_or_default())
        }
        async fn find_by_id(&self, id: u32) -> Result<Option<Staff>, DomainError> {
            Ok(Some(make_staff(id, false)))
        }
        async fn find_by_provider(&self, _: i32, _: &str) -> Result<Option<Staff>, DomainError> {
            Ok(None)
        }
        async fn find_all_active(&self) -> Result<Vec<Staff>, DomainError> {
            Ok(vec![])
        }
        async fn save(&self, s: Staff) -> Result<Staff, DomainError> {
            Ok(s)
        }
        async fn update_role(&self, _: u32, _: i32, _: u32) -> Result<bool, DomainError> {
            Ok(*self.update_role_ok.lock().unwrap())
        }
        async fn soft_delete(&self, _: u32, _: u32) -> Result<bool, DomainError> {
            Ok(*self.soft_delete_ok.lock().unwrap())
        }
        async fn restore(&self, _: u32) -> Result<bool, DomainError> {
            Ok(*self.restore_ok.lock().unwrap())
        }
    }

    #[tokio::test]
    async fn test_find_by_condition_maps_status() {
        let mock = Arc::new(MockRepo::new());
        *mock.find_by_condition.lock().unwrap() = Some(vec![
            make_staff(1, false),
            make_staff(2, true),
        ]);
        let uc = Interactor::new(mock);
        let result = uc.find_by_condition(Condition { keyword: None, roles: vec![] }).await.unwrap();
        assert_eq!(result.len(), 2);
        assert_eq!(result[0].status, 1);
        assert_eq!(result[1].status, 0);
    }

    #[tokio::test]
    async fn test_update_role_success() {
        let mock = Arc::new(MockRepo::new());
        let uc = Interactor::new(mock);
        let dto = UpdateRoleDto { id: 1, role: 1, executor_id: 99 };
        assert!(uc.update_role(dto).await.is_ok());
    }

    #[tokio::test]
    async fn test_update_role_not_found() {
        let mock = Arc::new(MockRepo::new());
        *mock.update_role_ok.lock().unwrap() = false;
        let uc = Interactor::new(mock);
        let dto = UpdateRoleDto { id: 999, role: 1, executor_id: 99 };
        assert!(uc.update_role(dto).await.is_err());
    }

    #[tokio::test]
    async fn test_destroy_success() {
        let mock = Arc::new(MockRepo::new());
        let uc = Interactor::new(mock);
        let dto = DestroyDto { id: 1, executor_id: 99 };
        assert!(uc.destroy(dto).await.is_ok());
    }

    #[tokio::test]
    async fn test_restore_not_found() {
        let mock = Arc::new(MockRepo::new());
        *mock.restore_ok.lock().unwrap() = false;
        let uc = Interactor::new(mock);
        assert!(uc.restore(999).await.is_err());
    }
}

File: backends/go-beego/internal/usecase/staff/interactor.go
// Package staff はスタッフユースケースを提供します。
package staff

import (
	domstaff "authorization-go-beego/internal/domain/staff"
	"authorization-go-beego/pkg/apperror"
)

// Interactor はスタッフユースケースの実装です。
type Interactor struct {
	repo domstaff.Repository
}

// NewInteractor は Interactor を生成します。
func NewInteractor(repo domstaff.Repository) *Interactor {
	return &Interactor{repo: repo}
}

// CountByCondition は条件に合うスタッフの総件数を返します。
func (uc *Interactor) CountByCondition(cond domstaff.Condition) (int, error) {
	return uc.repo.CountByCondition(cond)
}

// FindByCondition は条件に合うスタッフ一覧を取得します。
func (uc *Interactor) FindByCondition(cond domstaff.Condition) ([]*domstaff.ListItem, error) {
	staffs, err := uc.repo.FindByCondition(cond)
	if err != nil {
		return nil, err
	}
	items := make([]*domstaff.ListItem, 0, len(staffs))
	for _, s := range staffs {
		items = append(items, staffToListItem(s))
	}
	return items, nil
}

// UpdateRole はスタッフのロールを変更します。
func (uc *Interactor) UpdateRole(dto UpdateRoleDto) error {
	if dto.Role != domstaff.RoleAdmin && dto.Role != domstaff.RoleMember {
		return apperror.BadRequest("role_invalid")
	}
	s, err := uc.repo.FindByID(&domstaff.Staff{ID: dto.ID})
	if err != nil {
		return err
	}
	if s == nil {
		return apperror.NotFound("staff_not_found")
	}
	ok, err := uc.repo.UpdateRole(&domstaff.Staff{ID: dto.ID, Role: dto.Role, UpdatedBy: &dto.ExecutorID, Version: s.Version})
	if err != nil {
		return err
	}
	if !ok {
		return apperror.NotFound("staff_not_found")
	}
	return nil
}

// Destroy はスタッフを論理削除します。
func (uc *Interactor) Destroy(dto DestroyDto) error {
	s, err := uc.repo.FindByID(&domstaff.Staff{ID: dto.ID})
	if err != nil {
		return err
	}
	if s == nil {
		return apperror.NotFound("staff_not_found")
	}
	ok, err := uc.repo.SoftDelete(&domstaff.Staff{ID: dto.ID, DeletedBy: &dto.ExecutorID, Version: s.Version})
	if err != nil {
		return err
	}
	if !ok {
		return apperror.NotFound("staff_not_found")
	}
	return nil
}

// Restore は論理削除したスタッフを復元します。
func (uc *Interactor) Restore(dto RestoreDto) error {
	ok, err := uc.repo.Restore(&domstaff.Staff{ID: dto.ID})
	if err != nil {
		return err
	}
	if !ok {
		return apperror.NotFound("staff_not_found")
	}
	return nil
}

// staffToListItem はスタッフエンティティを一覧用VOに変換します。
func staffToListItem(s *domstaff.Staff) *domstaff.ListItem {
	return &domstaff.ListItem{
		ID:        s.ID,
		Name:      s.Name,
		Email:     s.Email,
		Role:      s.Role,
		Status:    staffStatus(s),
		CreatedAt: s.CreatedAt,
		UpdatedAt: s.UpdatedAt,
	}
}

// staffStatus はスタッフの有効/削除状態を返します。
func staffStatus(s *domstaff.Staff) int {
	if s.DeletedAt != nil {
		return 0
	}
	return 1
}
